### PR TITLE
fix: report new window downloads

### PR DESCRIPTION
**Pull request by yury-s**

### CODE REVIEW

File: browsers.json [hunk 1]
```diff
@@ -6,7 +6,7 @@
     },
     {
       "name": "firefox",
-      "revision": "1087"
+      "revision": "1088"
```
Comment (pavelfeldman) [inline browsers.json:9]: run `npm run doc` to generate readme.

File: src/webkit/wkBrowser.ts [hunk 2]
```diff
@@ -107,7 +107,13 @@ export class WKBrowser extends BrowserBase {
       // here by simulating cancelled provisional load which matches downloads from network.
       frameManager.provisionalLoadFailed(frame, '', 'Download is starting');
     }
-    this._downloadCreated(page._page, payload.uuid, payload.url);
+    let originPage = page._initializedPage;
+    // If it's a new window download, report it on the opener page.
+    if (!originPage)
+      originPage = await page.opener();
```
Comment (pavelfeldman) [inline src/webkit/wkBrowser.ts:113]: same here

File: src/firefox/ffBrowser.ts [hunk 3]
```diff
@@ -148,12 +148,22 @@ export class FFBrowser extends BrowserBase {
...
+  async _onDownloadCreated(payload: Protocol.Browser.downloadCreatedPayload) {
     const ffPage = this._ffPages.get(payload.pageTargetId)!;
     assert(ffPage);
     if (!ffPage)
       return;
-    this._downloadCreated(ffPage._page, payload.uuid, payload.url);
+    let originPage = ffPage._initializedPage;
+    // If it's a new window download, report it on the opener page.
+    if (!originPage) {
+      originPage = await ffPage.opener();
```
Comment (pavelfeldman) [inline src/firefox/ffBrowser.ts:159]: same here

File: src/chromium/crPage.ts [hunk 4]
```diff
@@ -649,8 +649,18 @@ class FrameSession {
     this._page._onFileChooserOpened(handle);
   }
 
-  _onDownloadWillBegin(payload: Protocol.Page.downloadWillBeginPayload) {
-    this._crPage._browserContext._browser._downloadCreated(this._page, payload.guid, payload.url);
+  async _onDownloadWillBegin(payload: Protocol.Page.downloadWillBeginPayload) {
+    let originPage = this._crPage._initializedPage;
+    // If it's a new window download, report it on the opener page.
+    if (!originPage) {
+      originPage = await this._crPage.opener();
```
Comment (pavelfeldman) [inline src/chromium/crPage.ts:656]: this now races with the download finished, i.e. crashes

File: browsers.json (hunk 1 above)
Comment (yury-s) [reply browsers.json:9]: Done. I don't see any diff though.

File: src/webkit/wkBrowser.ts (hunk 2 above)
Comment (yury-s) [reply src/webkit/wkBrowser.ts]: Done.

File: src/firefox/ffBrowser.ts (hunk 3 above)
Comment (yury-s) [reply src/firefox/ffBrowser.ts]: Done.

File: src/chromium/crPage.ts (hunk 4 above)
Comment (yury-s) [reply src/chromium/crPage.ts]: Done.

**Review by pavelfeldman — APPROVED** [review]

**Pull request merged**